Add feedback session button is not disabled for instructors without permissions for any course #3712

kennho opened the pull request:
Fixes #3712

thyageshm commented:
@yj-soh ready for review

yj-soh commented on src/main/java/teammates/ui/template/FeedbackSessionsForm.java:
```diff
@@ -156,7 +156,7 @@ public static FeedbackSessionsForm getFormForNewFs(FeedbackSessionAttributes fee
                                                                               : feedbackSession.gracePeriod);
         
-        newFsForm.isSubmitButtonDisabled= courseIds.isEmpty();
+        newFsForm.isSubmitButtonDisabled = courseIdOptions.get(0).getContent().equals("No active courses!");
```
Is there no better way than a brittle string comparison?
How about ensuring `courseIds` is indeed empty when there are no active courses or even shifting part of the logic for having no active courses into this method from `InstructorFeedbacksPageData`?

What exactly is an "active course"?
Does the instructor need permission to modify its session for it to be active for that instructor?
What should be the value of `isShowNoCoursesMessage` here in this case?

kennho commented:
Updated @yj-soh. As for the `isShowNoCoursesMessage`, it is to be displayed when the instructor have no courses, not when the instructor have no permission to edit courses. 

I have updated the dropdown to show a more appropriate message ("No permission to modify courses!") rather than ("No active courses!") as they actually meant two different things. However, I am not sure if this should be done here as it is already more than what the issue was opened for.

yj-soh commented:
Are you sure they do mean two different things?
I've just tried it, but on the current master, if you have a course and remove your own permission to modify its sessions, you do get the "No active courses" message in the select.
I'm not too sure about changing this present behaviour.

Any chance you can clarify on the definition of an "active course" for an instructor, @kanghj, @damithc ?

kennho commented:
Hmm in order to see it, you must remove your own permission to modify session for **all** of the courses that you have. This way, you would still have courses under you but you can't modify any of them. The `isShowNoCoursesMessage` won't be shown too because you still have courses, just that you can't modify them. But the "No active courses!" in the select will be shown. 

The "No active courses!" is still shown as currently the logic to show that only check if the select is empty then append the "No active courses!". It doesn't check why it is empty. Is it because of the instructor doesn't have any course, or is it because the instructor doesn't have permission to modify for all the courses that he have.

After you delete all of the courses under you, the 'isShowNoCoursesMessage' will then be shown, which is "You need to have an active(unarchived) course to create a session!" This message sounds like it is intended to be shown to instructor without any courses rather than instructor with courses but have no permission for any. Therefore the "active courses" seem to be saying the instructor have no courses rather than have no permissions to modify for all of his courses. @yj-soh 

![screen shot 2016-03-27 at 1 33 23 pm](https://cloud.githubusercontent.com/assets/10682388/14063992/98509db4-f420-11e5-9ec2-21dc2b31eade.png)

kanghj commented:
@yj-soh, I'm guessing "active" course just means unarchived

damithc commented:
Yes, 'no active courses' means all courses are archived. This is not the same as not having permissions to add sessions in any non-archived courses.

yj-soh commented on src/main/java/teammates/ui/controller/InstructorFeedbacksPageData.java:
```diff
@@ -271,15 +272,39 @@ public FeedbackSessionsCopyFromModal getCopyFromModal() {
+        if (result.isEmpty() && !courses.isEmpty()) {
+            defaultOption = createOption("No permission to modify courses!", "", true);
+            result.add(defaultOption);
+        } if (result.isEmpty()) {
```
The behaviour's probably the same, but is there supposed to be a newline or an `else` here?
That said, this would be best represented with a nested `if` for the courses empty check.

Also, if you do intend to have this other option, do provide a more accurate message.
It's the lack of permission to modify the courses' sessions, rather than the courses themselves.

I do wonder if it's better to use a single message for both cases, though, like maybe "No usable courses found!".
The instructor should have the same capabilities in either case.
This can be a new issue.

kennho replied:
Okays. So I will create a new issue for the lack of permission to modify the courses' sessions?

yj-soh replied:
A new issue for settling on a proper message in those cases?
What's wrong with the lack of permission to modify the courses' sessions, specifically?

yj-soh commented on src/main/java/teammates/ui/controller/InstructorFeedbacksPageData.java:
```diff
@@ -271,15 +272,39 @@ public FeedbackSessionsCopyFromModal getCopyFromModal() {
+    private boolean isNoModifyPermissionForAllCourses(List<CourseAttributes> courses, Map<String, InstructorAttributes> instructors) {
+        for (CourseAttributes course : courses) {
+            InstructorAttributes instructor = instructors.get(course.id);
+            boolean isAllowedForPriviledge = instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION);
```
Likewise, it's the sessions, not the course.

And seeing as this loop and conditional is already done at L253 - 268, shouldn't it be possible to re-use the result somehow?
Maybe calculate this value closer to the start before using it for both purposes?
I would honestly recommend just extracting the `defaultOption` segment into another method and using it from `buildBasicForm()` or something.

kennho commented:
Updated with the suggested changes. Is this okay? @yj-soh

kennho commented:
I have also updated this to combine all the merge @yj-soh

yj-soh commented on src/main/java/teammates/ui/controller/InstructorFeedbacksPageData.java:
```diff
@@ -270,15 +284,18 @@ public FeedbackSessionsCopyFromModal getCopyFromModal() {
+    private void addDefaultOption(List<ElementTag> selectOptions) {
+        ElementTag defaultOption = createOption("No active courses!", "", true);
+        selectOptions.add(defaultOption);
+    }
```
Try changing the "default" term used here.
"default" doesn't seem very suitable as it somewhat implies that it's an actual option.
Maybe something like "placeholder" instead?

kennho commented:
Sorry for slow reply. I am doing my other assignment. Will take a look later after I have completed my assignment. @yj-soh

kennho commented:
Updated with suggested changes @yj-soh

kennho commented:
@yj-soh this too. I really need it to be approved and merge soon
